src/rust/ide/view/src/searcher/icons.rs

```diff
@@ -0,0 +1,387 @@
+            // rotations to form the whole star.
+            let fragment1 = spike + pentagon_fragment;
+            let fragment2 = fragment1.rotate((rotation_step * 1.0).radians());
+            let fragment3 = fragment1.rotate((rotation_step * 2.0).radians());
+            let fragment4 = fragment1.rotate((rotation_step * 3.0).radians());
+            let fragment5 = fragment1.rotate((rotation_step * 4.0).radians());
+
+            let shape = fragment1 + fragment2 + fragment3 + fragment4 + fragment5;
+            let shape = shape.fill(style.get_color(theme::favorites));
+            let shape = shape.translate_y((-0.5).px());
+            shape.shrink(SHRINK_AMOUNT.px()).into()
+        }
+    }
+}
+
+
+
+// ==========
+// === IO ===
+// ==========
+
+// A rounded rectangle with an arrow pointing in from the left.
+mod data_input {
+    use super::*;
+
+    ensogl::define_shape_system! {
+        (style:Style) {
+
+            // === Rectangle ===
+
+            let rect = Rect((10.0.px(),13.0.px())).corners_radius(1.5.px()).translate_x(1.5.px());
+            // Taking just an outline.
+            let rect = &rect - rect.shrink(1.0.px());
+            // Creating a gap for the arrow to pass through.
+            let gap  = Rect((2.0.px(),4.0.px())).translate_x((-3.0).px());
+            let rect = rect - gap;
+
+
+            // === Arrow ===
+
+            let arrow_line = Rect((8.0.px(),1.0.px())).translate_x((-3.0).px());
+            let arrow_tip  = Triangle(5.0.px(),4.0.px()).rotate((PI / 2.0).radians());
+            let arrow_tip  = arrow_tip.translate_x(2.0.px());
+            let arrow      = arrow_line + arrow_tip;
+
+
+            // === Shape ===
+
+            let shape = rect + arrow;
+            let shape = shape.fill(style.get_color(theme::io));
+            shape.shrink(SHRINK_AMOUNT.px()).into()
+        }
+    }
+}
+
+// A rounded rectangle with an arrow pointing out to the right.
+mod data_output {
+    use super::*;
+
+    ensogl::define_shape_system! {
+        (style:Style) {
+
+            // === Rectangle ===
+            let rect = Rect((9.0.px(),13.0.px())).corners_radius(1.5.px()).translate_x((-2.5).px());
+            // Taking just an outline.
+            let rect = &rect - rect.shrink(1.0.px());
+            // Creating a gap for the arrow to pass through.
+            let gap  = Rect((2.0.px(),4.0.px())).translate_x((1.5).px());
+            let rect = rect - gap;
+
+
+            // === Arrow ===
+
+            let arrow_line = Rect((8.0.px(),1.0.px())).translate_x(1.0.px());
+            let arrow_tip  = Triangle(5.0.px(),4.0.px()).rotate((PI / 2.0).radians());
+            let arrow_tip  = arrow_tip.translate_x(6.0.px());
+            let arrow      = arrow_line + arrow_tip;
+
+
+            // === Shape ===
+
+            let shape = rect + arrow;
+            let shape = shape.fill(style.get_color(theme::io));
+            shape.shrink(SHRINK_AMOUNT.px()).into()
+        }
+    }
+}
```

Comment:
This has many common things with data input. Are we able to extract them?

Reply:
Can you tell me, what should be the return type of such a function?

Reply:
You can always click RMB in intellij and expand the macro - then you'll get the types. This is something like `AnyShape` or something like that.

Reply:
Which macro should I expand? I can take the lines from the shape system definition and move them into a function. Then the return type will be something like `Union<Translate<Triangle>,...>`. But this type contains so much irrelevant information. I would rather use `impl ...` as return type, or any other solution that doesn't require me to have that full type in the code.